src/System.Private.CoreLib/src/System/Threading/Thread.CoreRT.cs

```diff
@@ -327,14 +327,121 @@ public bool Join(int millisecondsTimeout)
+                }
+
+                return GetOptimalMaxSpinWaitsPerSpinIteration();
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static int GetOptimalMaxSpinWaitsPerSpinIteration()
+        {
+            InitializeYieldProcessorNormalized();
+            return Volatile.Read(ref s_optimalMaxNormalizedYieldsPerSpinIteration);
+        }
 
-        public static void SpinWait(int iterations) => RuntimeImports.RhSpinWait(iterations);
+        private static int s_optimalMaxNormalizedYieldsPerSpinIteration;
+        private static int s_yieldsPerNormalizedYield;
+
+        private static bool IsYieldProcessorNormalizedInitialized()
+        {
+            return s_yieldsPerNormalizedYield != 0 && s_optimalMaxNormalizedYieldsPerSpinIteration != 0;
+        }
+
+        private static void EnsureYieldProcessorNormalizedInitialized()
+        {
+            if (!IsYieldProcessorNormalizedInitialized())
+            {
+                InitializeYieldProcessorNormalized();
+            }
+        }
+
+        private static void InitializeYieldProcessorNormalized()
+        {
+            // TODO: critical section so that it only initializes once
+
+            // Intel pre-Skylake processor: measured typically 14-17 cycles per yield
+            // Intel post-Skylake processor: measured typically 125-150 cycles per yield
+            const int DefaultYieldsPerNormalizedYield = 1; // defaults are for when no measurement is done
+            const int DefaultOptimalMaxNormalizedYieldsPerSpinIteration = 64; // tuned for pre-Skylake processors, for post-Skylake it should be 7
+            const int MeasureDurationMs = 10;
+            const int MaxYieldsPerNormalizedYield = 10; // measured typically 8-9 on pre-Skylake
+            const int MinNsPerNormalizedYield = 37; // measured typically 37-46 on post-Skylake
+            const int NsPerOptimialMaxSpinIterationDuration = 272; // approx. 900 cycles, measured 281 on pre-Skylake, 263 on post-Skylake
+            const int NsPerSecond = 1000 * 1000 * 1000;
+
+            ulong ticksPerSecond = HighPerformanceCounter.Frequency;
+            if (ticksPerSecond < 1000 / MeasureDurationMs)
+            {
+                // High precision clock not available or clock resolution is too low, resort to defaults
+                s_yieldsPerNormalizedYield = DefaultYieldsPerNormalizedYield;
+                s_optimalMaxNormalizedYieldsPerSpinIteration = DefaultOptimalMaxNormalizedYieldsPerSpinIteration;
+                return;
+            }
+
+            // Measure the nanosecond delay per yield
+            ulong measureDurationTicks = ticksPerSecond / (1000 / MeasureDurationMs);
+            uint yieldCount = 0;
+            ulong startTicks = HighPerformanceCounter.TickCount;
+            ulong elapsedTicks;
+            do
+            {
+                RuntimeImports.RhSpinWait(10);
```

Comment:
10 is very short. Not sure what the implementation of `HighPerformanceCounter.TickCount` is on Unixes, but in CoreCLR it is significantly slower than on Windows and probably takes longer than 10 spin-waits on pre-Skylake processor, so the measurement may be off. I would recommend using the same value as in CoreCLR, 1000.